docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+## Commit-Log based Trillian storage
+
+*Status: Draft*
+
+*Authors: al@google.com, drysdale@google.com, filippo@cloudflare.com*
+
+*Last Updated: 2017-05-11*
+
+
+[TOC]
+
+
+## Objective
+
+A design for an alternative Trillian storage layer which uses a distributed and
+immutable *commit log* as the source of truth for a Trillian Log's contents and
+sequence information, and one or more independent *"readonly"* databases built
+from the commit log to serve queries.
+
+This design allows for:
+
+*   flexibility in scaling Trillian deployments,
+*   easier recovery from corrupt/failed database deployments since operators
+can simply delete it and allow it to rebuild itself from the commit log.
+
+Initially, this will be built using Apache Kafka for the commit log, with
+Apache HBase instances for the serving databases, since this is what Cloudflare
+has operational experience in running, but other distributed commit-log and
+database engines may be available.
+
+Having Trillian support a commit-log based storage system too will ensure
+Trillian doesn't inadvertently tie itself exclusively to strong globally
+consistent storage.
+
+## Background
+
+Trillian currently supports two storage technologies, mysql and Spanner, which
+provide strong global consistency.
+
+The design presented here requires:
+
+*   A durable, ordered, and immutable commit log.
+*   A "local" storage mechanism which can support the operations required by
+    the Trillian log_storage API.
+
+
+
+## Design Overview
+
+![overview diagram](commit_log_based_storage_design_overview.png "Overview")
+
+The `leaves` topic is the canonical source of truth for the ordering of leaves
+in a log.
+
+The `STHs` topic is a list of all STHs for a given log.
+
+Kafka topics are configured never to expire entries (this is a supported mode),
+and Kafka is known to scale to multiple terabytes within a single partition.
+
+HBase instances are one-per-cluster, built from the contents of the Kafka
+topics, and, consequently, are essentially disposable.
+
+Queued leaves are sent by the Trillian frontends to the Kafka `Leaves` topic.
+Since Kafka topics are append-only and immutable, this effectively sequences
+the entries in the queue.
+The current master sequencer consumes the `Leaves` topic, uses the Kafka sequence
+number of each dequeued leaf as its Trillian Log sequence number, and adds it to
+the local Merkle tree ensuring the sequence numbers match.
+
+When a new tree head is created, the sequencer writes this to the STHs topic
```

Comment:
I've reworded this a bit, I think it was slightly misleading before.
Have another look and see what you think.